src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buf.gradle
+
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputFiles
+import java.io.File
+import java.nio.charset.StandardCharsets
+
+/**
+ * A task executing buf executable as part of its operation.
+ */
+abstract class AbstractBufExecTask : AbstractBufTask() {
+    /** The buf executable. */
+    @get:InputFiles
+    abstract val bufExecutable: ConfigurableFileCollection
+
+    /**
+     * The directory in which buf is executed.
+     * The actual real input files in this directory have to be tracked per command separately,
+     * so it is just an @Input, not @InputDirectory.
+     */
+    @get:Input
+    abstract val workingDir: Property<File>
+
+    /** Whether the project has protobuf plugin enabled. */
+    @get:Input
+    abstract val hasProtobufGradlePlugin: Property<Boolean>
+
+    /** Directories possibly containing input .proto files. */
+    @get:InputFiles
+    abstract val candidateProtoDirs: ConfigurableFileCollection
+
+    /** Whether the project has buf workspace or not. */
+    @get:Input
+    abstract val hasWorkspace: Property<Boolean>
+
+    internal fun execBuf(
```

Comment:
I think I'd prefer keeping these utility functions in their separate files as extension functions. They don't really fit the "traditional" abstract class pattern of, for example, calling into overrideable methods, and I think it's nice to keep the abstract classes as simple containers of properties.

Reply:
I've reworked all commits to comply to this rule (and to make as minimal changes in the structure as possible).

Reply:
Looks great - thanks!